tratto/src/main/java/star/tratto/util/javaparser/JavaParserUtils.java

```diff
@@ -0,0 +1,1078 @@
+                " " + getTypeWithoutPackages(methodDeclaration.getReturnType()) +
+                " " + methodDeclaration.getName() +
+                "(" + String.join(", ", formalParameterList) + ")" +
+                (exceptionList.isEmpty() ? "" : " throws " + String.join(", ", exceptionList)))
+                .replaceAll(" +", " ").trim();
+    }
+
+    /**
+     * Gets the JavaParser package declaration of a given compilation unit.
+     *
+     * @param cu a compilation unit
+     * @return the JavaParser package declaration {@link PackageDeclaration}
+     * of the compilation unit
+     * @throws PackageDeclarationNotFoundException if the JavaParser package
+     * declaration {@link PackageDeclaration} cannot be found or is unnamed
+     */
+    public static PackageDeclaration getPackageDeclaration(
+            CompilationUnit cu
+    ) throws PackageDeclarationNotFoundException {
+        Optional<PackageDeclaration> jpPackage = cu.getPackageDeclaration();
+        if (jpPackage.isEmpty()) {
+            throw new PackageDeclarationNotFoundException("The compilation unit has no package declaration");
+        }
+        return jpPackage.get();
+    }
+
+    /**
+     * Returns the base component type of a resolved type. Recursively strips
+     * all array variables. For example:
+     *     Object[][] => Object
+     *
+     * @param resolvedType a type
+     * @return the base component type
+     */
+    public static ResolvedType removeArray(ResolvedType resolvedType) {
+        while (resolvedType.isArray()) {
+            resolvedType = resolvedType.asArrayType().getComponentType();
+        }
+        return resolvedType;
+    }
+
+    /**
+     * Checks if a type is a type variable.
+     *
+     * @param resolvedType a JavaParser resolved type
+     * @return true iff a given type is a type variable of a generic class or
+     * method. If the given type is an array, then this method checks the base
+     * component type.
+     */
+    public static boolean isTypeVariable(
+            ResolvedType resolvedType
+    ) {
+        ResolvedType componentType = removeArray(resolvedType);
+        return componentType.isTypeVariable();
+    }
+
+    /**
+     * Checks if a type name represents a type variable declared in the given
+     * class or method. This method checks both the method where the type is
+     * used and its declaring class.
+     *
+     * @param typeName a type name
+     * @param jpCallable the method using {@code typeName}
+     * @param jpClass the declaring class of {@code jpCallable}
+     * @return true iff a given type is a type variable of the declaring class
+     * or method. If the given type is an array, then this method checks the
+     * base component type.
+     */
+    public static boolean isTypeVariable(
+            String typeName,
+            CallableDeclaration<?> jpCallable,
+            TypeDeclaration<?> jpClass
+    ) {
+        // add all type parameters of the method
```

Comment:
This implementation is quite inefficient.  It doesn't need to create a list, and it can exit early.  Have two for loops, with a test and `return true` in the body of each.